### PR TITLE
Proposed fix for #3529 (Functionality decoupling for _strat_commutes_…

### PR DESCRIPTION
**Pull request by akushnarov**
1. Moved "commute" check from `protocols/commutes_protocol.py` to `ops/raw_types.py:Operator`
2. Adjusted the "ops/gate_operation.py" (extended `_commute_` functionality to check if the parent class commutes)

### CODE REVIEW

File: cirq/protocols/commutes_protocol_test.py [hunk 1]
```diff
@@ -15,6 +15,7 @@
 import numpy as np
 import pytest
 import sympy
+import copy
```
Comment (vtomole) [inline cirq/protocols/commutes_protocol_test.py:18]: You are not using this import. Remove it.

File: cirq/protocols/commutes_protocol.py [hunk 2]
```diff
@@ -185,31 +184,4 @@ def _strat_commutes_from_matrix(
...
-    # Don't create gigantic matrices.
-    if np.product(qid_shape_protocol.qid_shape(circuit12)) > 2**10:
-        return NotImplemented  # coverage: ignore
-
-    m12 = unitary_protocol.unitary(circuit12, default=None)
-    m21 = unitary_protocol.unitary(circuit21, default=None)
-    if m12 is None:
-        return NotImplemented
-    return np.allclose(m12, m21, atol=atol)
+    return linalg.matrix_commutes(v1, v2, atol=atol)
```
Comment (vtomole) [inline cirq/protocols/commutes_protocol.py:187]: Add newline at the end of this file.

File: cirq/ops/pauli_string.py [hunk 3]
```diff
@@ -652,8 +652,8 @@ def zip_paulis(self, other: 'cirq.PauliString'
                   ) -> Iterator[Tuple[pauli_gates.Pauli, pauli_gates.Pauli]]:
         return (paulis for qubit, paulis in self.zip_items(other))
 
-    def _commutes_(self, other: Any,
-                   atol: float) -> Union[bool, NotImplementedType, None]:
+    def _commutes_(self, other: Any, *, atol: Union[int, float] = 1e-8
```
Comment (vtomole) [inline cirq/ops/pauli_string.py:655]: Why the addition of ` 1e-8`?

**Review by vtomole — CHANGES_REQUESTED** [review]

File: cirq/protocols/commutes_protocol_test.py (hunk 1 above)
Comment (akushnarov) [reply cirq/protocols/commutes_protocol_test.py:18]: Roger that

File: cirq/protocols/commutes_protocol.py (hunk 2 above)
Comment (akushnarov) [reply cirq/protocols/commutes_protocol.py:187]: This comment is a bit confusing since this code (lines 213:215 from the comment) was moved to ops/raw_types.py:Operation:_commutes_

File: cirq/ops/pauli_string.py (hunk 3 above)
Comment (akushnarov) [reply cirq/ops/pauli_string.py:655]: I've just added a default value. Across Cirq code there are sometimes `_commutes_` methods with the default value and sometimes without, e.g.:
- cirq/ops/raw_types.py:Gate -> `def _commutes_(self, other: Any, atol: float) -> Union[None, NotImplementedType, bool]:`
- cirq/ops/raw_types.py:TaggedOperation -> `def _commutes_(self, other: Any, *, atol: Union[int, float] = 1e-8) -> Union[bool, NotImplementedType, None]:`

What are the advantages of removing the default value?

File: cirq/ops/raw_types.py [hunk 4]
```diff
@@ -507,6 +507,36 @@ def validate_args(self, qubits: Sequence['cirq.Qid']):
         """
         _validate_qid_shape(self, qubits)
 
+    def _commutes_(
+        self, other: Any, *, atol: Union[int, float] = 1e-8
+    ) -> Union[bool, NotImplementedType, None]:
+        """Determine if this Operation commutes with the object"""
+        if isinstance(other, str):
```
Comment (balopat) [inline cirq/ops/raw_types.py:514]: Why aren't we checking for `other` to be the instance of `ops.Operation`? That was the original check as well.

File: cirq/ops/pauli_string.py [hunk 5]
```diff
@@ -698,7 +698,10 @@ def zip_paulis(
     ) -> Iterator[Tuple[pauli_gates.Pauli, pauli_gates.Pauli]]:
         return (paulis for qubit, paulis in self.zip_items(other))
 
-    def _commutes_(self, other: Any, atol: float) -> Union[bool, NotImplementedType, None]:
+    def _commutes_(
```
Comment (balopat) [inline cirq/ops/pauli_string.py:701]: this seems unrelated - can you revert this formatting change?

File: cirq/ops/gate_operation.py [hunk 6]
```diff
@@ -167,8 +167,18 @@ def _unitary_(self) -> Union[np.ndarray, NotImplementedType]:
...
-        return self.gate._commutes_on_qids_(self.qubits, other, atol=atol)
+    def _commutes_(
+        self, other: Any, atol: Union[int, float] = 1e-8
+    ) -> Union[bool, NotImplementedType, None]:
+
+        commutes = self.gate._commutes_on_qids_(self.qubits, other, atol=atol)
+        if commutes is not NotImplemented:
+            return commutes
+
+        getter = getattr(super(), '_commutes_', None)
```
Comment (balopat) [inline cirq/ops/gate_operation.py:178]: I have two questions: 
1. What motivated the changes in this method? 
2. Also, why do we need to check for the existence of `_commutes_` on `super()`? We _know_ that `cirq.Operation` implements `_commutes_` now, as we just implemented it...

File: cirq/ops/raw_types.py (hunk 4 above)
Comment (akushnarov) [reply cirq/ops/raw_types.py:514]: 1. In the prev implementation (with the strategy) we are checking that at least one of the arguments is Operation (because it might be anything when you pass to cirq.commutes(v1, v2))
2. In new implementation (a method `_commutes_` of the Operation class) we already know that `self` is an instance of the Operation class and `other` might be Operation or Gate or etc.

**Comment by akushnarov** [comment]
@balopat please see my answers above, also I adjusted the code

File: cirq/ops/raw_types.py (hunk 4 above)
Comment (balopat) [reply cirq/ops/raw_types.py:514]: The original logic was this: 

```
if not isinstance(v1, ops.Operation) or not isinstance(v2, ops.Operation):	
        return NotImplemented
```

This means that both values should be `Operation` otherwise it is not `NotImplemented`.

Comment (akushnarov) [reply cirq/ops/raw_types.py]: Oh, sorry, you are 100% right, my bad.  Fixed in the last commit

**Comment by akushnarov** [comment]
Hi @balopat, I'm back to Cirq. Could you please check the changes?

File: cirq/ops/pauli_string.py (hunk 5 above)
Comment (balopat) [reply cirq/ops/pauli_string.py:702]: in fact there seems to be some inconsistency around this in the codebase:

image

Alright, let's keep the * in front of atol, I'll open an issue to cleanup this inconsistency.

File: cirq/ops/gate_operation.py [hunk 7]
```diff
@@ -167,8 +167,14 @@ def _unitary_(self) -> Union[np.ndarray, NotImplementedType]:
             return getter()
         return NotImplemented
 
-    def _commutes_(self, other: Any, atol: float) -> Union[bool, NotImplementedType, None]:
-        return self.gate._commutes_on_qids_(self.qubits, other, atol=atol)
+    def _commutes_(
```
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/ops/gate_operation.py:170]: OK, I'll merge this, and then we can fix the inconsistency in #3695

**Review by balopat — APPROVED** [review]
